library/alloc/src/vec/mod.rs

```diff
@@ -1231,6 +1241,19 @@ impl<T, A: Allocator> Vec<T, A> {
     /// }
     /// ```
     ///
+    /// Due to the aliasing guarantee, the following code is legal:
+    ///
+    /// ```rust
+    /// unsafe {
+    ///     let mut v = vec![0];
+    ///     let ptr1 = v.as_ptr();
+    ///     let _ = ptr1.read();
+    ///     let ptr2 = v.as_mut_ptr();
+    ///     ptr2.write(2);
+    ///     // Notably, the write to `ptr2` did *not* invalidate `ptr1`:
+    ///     let _ = ptr1.read();
```

Comment:
The shared reference is for the Vec, though, and that's a short lived one validated by the borrow checker.

It's useful because what if you have an `as_ptr()` in one place and an `as_mut_ptr()` elsewhere, and they are reading from/writing to different parts of the vector? The typed-arena use case is very much like this.

Now, this example has a write to the _same_ place, but perhaps we should talk specifically about writes in the docs here and tweak the example to have non-overlapping reads/writes.

Reply:
> The shared reference is for the Vec, though,

Yes indeed. So do we want to commit to references only having "shallow" effects here? That is the question.

On an array, `as_ptr()` is very much a pointer that is not happy with conflicting writes. OTOH we already guarantee more than we can for arrays on `as_mut_ptr()`...

Reply:
Reading through the docs again, this is probably indeed the minimum to make a useful guarantee of "no reference to the slice is being created". Which is exactly how it differs from arrays.

Is it worth stating explicitly that this is a special property of `as_(mut_)ptr` on `Vec` and does not necessarily apply to other methods with the same name on different types?

Reply:
Yeah, probably!

> but we are saying that it tolerates writes through other pointers

we could say it tolerates writes to _other parts of the buffer_ through other pointers, though!

Reply:
> we could say it tolerates writes to other parts of the buffer through other pointers, though!

Yes that does not sound controversial at all.